File: lua/modifying_sequence.lua
-- copy if not existed
local function copy(src, dest)
    for k, v in pairs(src) do
        if dest[k] == nil then dest[k] = v end
    end
end

-- copy if the value in src is ok
local function copyIf(src, dest, func)
    for k, v in pairs(src) do
        if func(v) and dest[k] == nil then dest[k] = v end
    end
end

local function copyN(dest, n, value)
    for i = 1, n do dest[i] = value end
end

-- return the index copy starts. If copy fails. return 0
local function copyBackward(src, dest)
    local gap = #dest - #src
    if gap < 0 then return 0 end
    for i = 1, #src do dest[i + gap] = src[i] end
    return gap + 1
end

local function fill(table, value)
    for k, _ in pairs(table) do
        table[k] = value
    end
end

local function fillN(seq, n, value)
    for i = 1, #seq do
        if i <= n then seq[i] = value end
    end
end

local function transform(table, func)
    for k, v in pairs(table) do
        table[k] = func(v)
    end
end

local function generate(table, func)
    for k, _ in pairs(table) do
        table[k] = func()
    end
end

local function generateN(seq, n, func)
    for i = 1, #seq do
        if i <= n then seq[i] = func() end
    end
end

local function removeIf(tb, func)
    for k, v in pairs(tb) do
        if func(k, v) then table.remove(tb, k) end
    end
end

local function remove(seq, value)
    removeIf(seq, function (k, v)
        return v == value
    end)
end

local function removeCopyIf(tb, func)
    local copy = {}
    for k, v in pairs(tb) do
        if func(k, v) then
            table.remove(tb, k)
            table.insert(copy, v)
        end
    end
    return copy
end

local function removeCopy(tb, value)
    return removeCopyIf(tb, function (k, v)
        return v == value
    end)
end

local function replaceIf(tb, func, after)
    for k, v in pairs(tb) do
        if func(k, v) then tb[k] = after end
    end
end

local function replace(tb, before, after)
    replaceIf(tb, function (k, v)
        return v == before
    end, after)
end

local function replaceCopyIf(tb, func, after)
    local copy = {}
    for k, v in pairs(tb) do
        if func(k, v) then
            copy[k] = after
        else
            copy[k] = v
        end
    end
    return copy
end

local function replaceCopy(tb, before, after)
    return replaceCopyIf(tb, function (k, v)
        return v == before
    end, after)
end

local function swapRange(a, b, startAt, endAt)
    if startAt < 0 or endAt > #a or endAt > #b then return end
    for i = startAt, endAt do
        a[i], b[i] = b[i], a[i]
    end
end

local function reverseRange(seq, startAt, endAt)
    for i = startAt, (endAt + startAt) / 2 do
        seq[i], seq[startAt+endAt-i] = seq[startAt+endAt-i], seq[i]
    end
end

local function reverse(seq)
    reverseRange(seq, 1, #seq)
end

local function reverseCopy(seq)
    local copy = {}
    for i = #seq, 1, -1 do
        copy[#seq-i+1] = seq[i]
    end
    return copy
end

local function shuffle(seq)
    for i = 1, #seq do
        local rd = math.floor(math.random() * i) + 1
        seq[rd], seq[i] = seq[i], seq[rd]
    end
end

-- assume 1 <= pivot <= #seq
local function rotate(seq, pivot)
    if #seq == 0 or pivot <= 1 then return end
    reverseRange(seq, 1, pivot-1)
    reverseRange(seq, pivot, #seq)
    reverseRange(seq, 1, #seq)
end

local function rotateCopy(seq, pivot)
    local copy = {}
    for i = pivot, #seq do
        table.insert(copy, seq[i])
    end
    for i = 1, pivot-1 do
        table.insert(copy, seq[i])
    end
    return copy
end

-- unique function only works on list-like table
local function unique(seq)
    local preValue = nil
    local i = 1
    for j = 1, #seq do
        if seq[j] ~= preValue then
            preValue = seq[j]
            seq[i] = preValue
            i = i + 1
        end
    end
    for j = i-1, #seq do
        table.remove(seq, j)
    end
end

local function uniqueCopy(seq)
    local copy = {}
    local preValue = nil
    local i = 1
    for j = 1, #seq do
        if seq[j] ~= preValue then
            preValue = seq[j]
            copy[i] = preValue
            i = i + 1
        end
    end
    return copy
end

return {
    copy = copy,
    copyIf = copyIf,
    copyN = copyN,
    copyBackward = copyBackward,
    -- move* is the same as copy*
    move = copy,
    moveBackward = copyBackward,
    fill = fill,
    fillN = fillN,
    transform = transform,
    generate = generate,
    generateN = generateN,
    remove = remove,
    removeIf = removeIf,
    removeCopy = removeCopy,
    removeCopyIf = removeCopyIf,
    replace = replace,
    replaceIf = replaceIf,
    replaceCopy = replaceCopy,
    replaceCopyIf = replaceCopyIf,
    swapRange = swapRange,
    reverse = reverse,
    reverseCopy = reverseCopy,
    shuffle = shuffle,
    rotate = rotate,
    rotateCopy = rotateCopy,
    unique = unique,
    uniqueCopy = uniqueCopy,
}

File: typescript/src/modifyingSequence.ts
let copy = <T>(src: any, dest: any) => {
    for (let i in src) {
        if (dest[i] == undefined) {
            dest[i] = src[i]
        }
    }
}

let copyIf = <T>(src: any, dest: any, func:(e: T)=>boolean) => {
    for (let i in src) {
        if (dest[i] == undefined && func(src[i])) {
            dest[i] = src[i]
        }
    }
}

let copyN = <T>(dest: T[], n: number, value: T) => {
    for (let i = 0; i < n; i++) {
        dest[i] = value
    }
}

let copyBackward = <T>(src: T[], dest: T[]): number => {
    const gap = dest.length - src.length
    if (gap < 0) {
        return -1
    }
    for (let i = 0; i < src.length; i++) {
        dest[i + gap] = src[i]
    }
    return gap
}

let fill = <T>(ary: T[], value: T) => {
    for (let i = 0; i < ary.length; i++) {
        ary[i] = value
    }
}

let fillN = <T>(ary: T[], n: number, value: T) => {
    for (let i = 0; i < n; i++) {
        ary[i] = value
    }
}

let transform = <T>(ary: T[], func:(a: T) => T) => (
    ary.forEach((e: T, idx: number) => (ary[idx] = func(e))))

let generate = <T>(ary: T[], g:() => T) => (
    // forEach doesn't walk through null element
    generateN(ary, ary.length, g))

let generateN = <T>(ary: T[], n: number, g:() => T) => {
    for (let i = 0; i < n; i++) {
        ary[i] = g()
    }
}

let isArray = <T>(coll: any): coll is Array<T> => {
    return coll instanceof Array
}

let _removeIf = <T>(coll: any, func: (e: T) => boolean, needCopy: boolean) => {
    if (isArray(coll)) {
        const ary = new Array()
        for (let e of (coll as Array<T>)) {
            if (func(e) == needCopy) {
                ary.push(e)
            }
        }
        if (needCopy) {
            return ary
        }
        while (coll.length != ary.length) {
            coll.shift()
        }
        for (let i = 0; i < ary.length; i++) {
            coll[i] = ary[i]
        }
    } else if (needCopy) {
        let copy = coll.constructor()
        for (let i in coll) {
            if (coll.hasOwnProperty(i) && func(coll[i])) {
                copy[i] = coll[i]
            }
        }
        return copy
    } else {
        for (let i in coll) {
            if (coll.hasOwnProperty(i) && func(coll[i])) {
                delete coll[i]
            }
        }
    }
}

let removeIf = (coll: any, func: (e: any) => boolean) => (
    _removeIf(coll, func, false))

let remove = (coll: any, value: any) => (
    _removeIf(coll, (e: any) => (e == value), false))

let removeCopyIf = (coll: any, func: (e: any) => boolean) => (
    _removeIf(coll, func, true))

let removeCopy = (coll: any, value: any) => (
    _removeIf(coll, (e: any) => (e == value), true))

let _replaceIf = <T>(coll: any, func: (e: T) => boolean, after: T, needCopy: boolean) => {
    if (isArray(coll)) {
        const ary = new Array()
        coll.forEach((e: T, idx: number) => {
            if (!func(e)) {
                if (needCopy) {
                    ary.push(e)
                }
            } else if (needCopy) {
                    ary.push(after)
            } else {
                coll[idx] = after
            }
        })
        if (needCopy) {
            return ary
        }
    } else if (needCopy) {
        let copy = coll.constructor()
        for (let i in coll) {
            if (coll.hasOwnProperty(i) && func(coll[i])) {
                copy[i] = after
            } else {
                copy[i] = coll[i]
            }
        }
        return copy
    } else {
        for (let i in coll) {
            if (coll.hasOwnProperty(i) && func(coll[i])) {
                coll[i] = after
            }
        }
    }
}

let replaceIf = (coll: any, func: (e: any) => boolean, after: any) => (
    _replaceIf(coll, func, after, false))

let replace = (coll: any, value: any, after: any) => (
    _replaceIf(coll, (e: any) => (e == value), after, false))

let replaceCopyIf = (coll: any, func: (e: any) => boolean, after: any) => (
    _replaceIf(coll, func, after, true))

let replaceCopy = (coll: any, value: any, after: any) => (
    _replaceIf(coll, (e: any) => (e == value), after, true))

let swapRange = <T>(ary1: T[], ary2: T[], start: number, end: number) => {
    if (start < 0 || end > Math.min(ary1.length, ary2.length)) {
        return
    }
    for (let i = start; i < end; i++) {
        [ary1[i], ary2[i]] = [ary2[i], ary1[i]]
    }
}

let reverse = <T>(ary: T[]) => {
    const middle = ary.length / 2
    for (let i = 0; i < middle; i++) {
        [ary[i], ary[ary.length-i-1]] = [ary[ary.length-i-1], ary[i]]
    }
}

let reverseCopy = <T>(ary: T[]) => {
    let copy = new Array<T>(ary.length)
    for (let i = ary.length - 1; i >= 0; i--) {
        copy[ary.length - i - 1] = ary[i]
    }
    return copy
}

let shuffle = <T>(ary: T[]) => {
    for (let i = 0; i < ary.length; i++) {
        const random : number = Math.floor(Math.random() * (i + 1));
        [ary[i], ary[random]] = [ary[random], ary[i]]
    }
}

let rotate = <T>(ary: T[], pivot: number) => (
    ary.push(...ary.splice(0, pivot)))

let rotateCopy = <T>(ary: T[], pivot: number) => {
    return ary.slice(pivot, ary.length).concat(ary.slice(0, pivot))
}

let _unique = <T>(ary: T[], needCopy: boolean) => {
    if (ary.length == 0) return
    let curValue = ary[0]
    let remain : T[] = [curValue]
    for (let i = 1; i < ary.length; i++) {
        if (ary[i] != curValue) {
            curValue = ary[i]
            remain.push(curValue)
        }
    }
    if (needCopy) {
        return remain
    }
    while (ary.length > remain.length) {
        ary.shift()
    }
    for (let i = 0; i < remain.length; i++) {
        ary[i] = remain[i]
    }
}

let unique = <T>(ary: T[]) => (_unique(ary, false))
let uniqueCopy = <T>(ary: T[]) => (_unique(ary, true))

export {
    copy,
    copyIf,
    copyN,
    copyBackward,
    fill,
    fillN,
    transform,
    generate,
    generateN,
    remove,
    removeIf,
    removeCopy,
    removeCopyIf,
    replace,
    replaceIf,
    replaceCopy,
    replaceCopyIf,
    swapRange,
    reverse,
    reverseCopy,
    shuffle,
    rotate,
    rotateCopy,
    unique,
    uniqueCopy,
}
